### PR TITLE
Dialogue-based language learning environment.

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
In this environment, the teacher verbally interacts with the learner about randomly selected object classes, and provides sentence feedback as well as reward feedback according to learner's response. This provides an environment for grounded language learning through
dialogue interactions.

### CODE REVIEW

File: examples/test_xworld_dialog.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+using namespace simulator::xwd;
+using namespace simulator;
+
+DECLARE_int32(context);
+DECLARE_bool(task_groups_exclusive);
+DECLARE_string(task_mode);
+DECLARE_bool(pause_screen);
+
+void game_reset_with_teacher(SimulatorPtr game, TeacherPtr teacher) {
```
Comment (yu239) [inline examples/test_xworld_dialog.cpp]: remove

Comment (skylian) [inline examples/test_xworld_dialog.cpp]: this function is not being used? plus xwd now has teacher inside

Comment (Haichao-Zhang) [reply examples/test_xworld_dialog.cpp]: Removed.

Comment (Haichao-Zhang) [reply examples/test_xworld_dialog.cpp]: Removed.

File: examples/test_xworld_dialog.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    xwd->reset_game();
+
+    auto num_actions = xwd->get_num_actions();
+    int act_rep = FLAGS_context;
+
+    double reward = 0;
+    double reward_per_game = 0;
+    double r = 0;
+    for (int i = 0; i < 100; i++) {
+        xwd->show_screen();
```
Comment (yu239) [inline examples/test_xworld_dialog.cpp]: comment this

File: games/xworld/maps/xworld_env.py [hunk 3]
```diff
@@ -54,13 +54,16 @@ class XWorldEnv(object):
 
     curriculum_check_period = 100
 
-    def __init__(self, item_path, max_height=7, max_width=7):
+    def __init__(self, item_path, max_height=7, max_width=7, class_per_session=-1):
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: don't put this in the base class; do this in your map file instead. See xworld3d/maps/XWorld3DNav.py

File: games/xworld/maps/xworld_env.py [hunk 4]
```diff
@@ -80,12 +84,23 @@ def __init__(self, item_path, max_height=7, max_width=7):
                             for l in lines if not l.startswith("//") and not l == ""}
 
     ############################ interface with Python tasks ############################
-    def reset(self):
+    def reset(self, same_session=False):
         """
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: Move this logic to your own map file; use derived class members to record the state of map changes.

Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: Done.

File: examples/test_xworld_dialog.cpp (hunk 2 above)
Comment (Haichao-Zhang) [reply examples/test_xworld_dialog.cpp]: Modified.

File: games/xworld/maps/xworld_env.py (hunk 3 above)
Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: Done.

File: games/xworld/maps/xworld_env.py [hunk 5]
```diff
@@ -138,18 +123,52 @@ def set_dims(self, h, w):
...
+            if not force_occupy:
+                assert loc in self.available_grids, \
+                    "set_dims correctly before setting a location"
+            if loc in self.available_grids:
+                self.available_grids.remove(loc)
         self.entity_nums[type] += 1
         self.entities.append(Entity(type=type, loc=loc, name=name))
         self.changed = True
 
+    def set_entity_instance(self, type, loc=None, name=None,
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: write a function to change the properties on the fly, without deleting and adding again.

File: games/xworld/maps/XWorldDialogMap.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    def select_goal_classes(self):
+        """
+        Sample a number of classes (class_per_session) for interaction within a session
+        """
+        if self.class_per_session > 1:
+            self.sel_classes = random.sample(self.items["goal"].keys(), self.class_per_session)
+        else:
+            self.sel_classes = self.items["goal"].keys()
+
+    def get_selected_classes(self):
```
Comment (skylian) [inline games/xworld/maps/XWorldDialogMap.py]: rename to get_selected_goal_classes?

Comment (Haichao-Zhang) [reply games/xworld/maps/XWorldDialogMap.py]: name changed.

File: games/xworld/maps/xworld_env.py (hunk 5 above)
Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: done.

File: games/xworld/maps/xworld_env.py [hunk 7]
```diff
@@ -123,18 +124,73 @@ def set_dims(self, h, w):
...
+                assert loc in self.available_grids, \
+                    "set_dims correctly before setting a location"
+            if loc in self.available_grids:
+                self.available_grids.remove(loc)
         self.entity_nums[type] += 1
         self.entities.append(Entity(type=type, loc=loc, name=name))
         self.changed = True
 
+    def reinstantiate_entity_property(self, entity, property=None, value=None):
+        """
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: Is it possible to replace the last two arguments with a dictionary where the key is the property? In this way, one call can change multiple propery-value pairs.

File: games/xworld/maps/xworld_env.py [hunk 8]
```diff
@@ -123,18 +124,73 @@ def set_dims(self, h, w):
...
+            entity.yaw = check_or_get_value(yaw_range) * self.PI_2
+        if (not property or property == "scale") \
+                                     and get_flag("visible_radius") and e.type == "goal":
+            scale_range = np.arange(0.5, 1.1, 0.1)
+            entity.scale = check_or_get_value(quantize_value(scale_range, 0.1))
+        if (not property or property == "offset") \
+                                     and get_flag("visible_radius") and e.type == "goal":
+            offset_range = np.arange(0, 1 - (entity.scale if hasattr(entity, 'scale') else 0.4), 0.1)
+            entity.offset = check_or_get_value(quantize_value(offset_range, 0.1))
+        # color is not exposed for reinstantiation
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: Why set the 0.1 as the step size for scale and offset?

Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: improved

File: games/xworld/maps/xworld_env.py (hunk 7 above)
Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: done

File: games/xworld/maps/xworld_env.py [hunk 9]
```diff
@@ -123,18 +124,91 @@ def set_dims(self, h, w):
...
+                entity.color = self.color_table[entity.asset_path]
+            if property == "yaw" and get_flag("visible_radius") and e.type != "block":
+                ## if partially observed, perturb the objects
+                yaw_range = range(-1, 3)
+                entity.yaw = check_or_get_value(yaw_range) * self.PI_2
+            if property == "scale" and get_flag("visible_radius") and e.type == "goal":
+                scale_range = [0.5, 1]
+                entity.scale = check_or_get_value(scale_range, is_continuous=True)
+            if property == "offset" and get_flag("visible_radius") and e.type == "goal":
+                offset_range = [0, 1 - (entity.scale if hasattr(entity, 'scale') else 0.5)]
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: Should assert hasattr(entity, 'scale'); otherwise offset cannot be decided.

**Comment by yu239** [comment]
squash and merge

**Pull request merged**